docs/prompting/ensembling/cosp.md

```diff
@@ -1,7 +1,267 @@
+2. We compute a score for each response using a weighted sum of two values - normalized entropy and repetitiveness ( How many times this rationale appears for this amswer )
+3. We rank all of our `nm` responses using this score and choose the `k` examples with the lowest scores as our final few shot examples.
+
+#### Normalized Entropy
+
+> In the paper, the authors write that normalized entropy is a good proxy over a number of different tasks where low entropy is positively correlated with correctness. Entropy is also supposed to range from 0 to 1.
+>
+> Therefore in order to do so, we introduce a `-` term in our implementation so that the calculated values range from 0 to 1.
+
+![](../../img/cosp_entropy.png)
+
+Assuming that for a specific question $x^{(i)}$, we have generated $m$ final answers of which $u$ are unique. ( Note that this only cares about the answer itself and not the rationale )
+
+$$
+\mathcal{H}\left(x^{(i)} \mid \left\{\hat{y}_j^{(i)}\right\}_{j=1}^m\right) = \frac{\sum_{\alpha=1}^u \hat{p}\left(\hat{y}_{\alpha}^{(i)}\right) \log \hat{p}\left(\hat{y}_{\alpha}^{(i)}\right)}{\log m},
+$$
+
+We can measure the entropy of the generated responses using the formula above where
+
+- $x_i$ is the original question that we prompted the model with
+- $y_j^{i}$ represents the $i$-th sampled response from the $m$ that we generated
+- $\hat{p}\left(\hat{y}_{\alpha}^{(i)}\right)$ is the frequency of the unique answer in all the $m$ generated answers. (Eg. if we generate 8 responses and 4 of them return the value 10, then $\hat{p}\left(\hat{y}_{\alpha}^{(i)}\right)$ is just going to be 0.5)
+
+#### Repetitiveness
+
+$$
+R_r(r_j^{(i)}) = \frac{2}{Q(Q-1)} \sum_{a=1}^{Q} \sum_{b=a+1}^{Q} W_{ab}
+$$
+
+In the formula above, $Q$ refers to the number of phrases in the sentence and $W_{ab}$ refers to the cosine similarity of two phrases $a$ and $b$.
+
+Repetitiveness aims to measure how often the language model repeats itself. To do so, the paper sums up the cosine similarity between each sentence inside the generated chain of thought rationale before normalizing it.
+
+The intuition behind this is that high repetitiveness indicates redundancy, which can lead to poorer performance. Therefore responses with a high number of similar sentences will have a larger score for repetitiveness ( since cosine similarity will be larger for each sentence ).
+
+### Step 2 - Self Consistency
+
+We now take our `k` responses and append them to our prompt. We then sample our model multiple times using this new prompt and take the majority vote as the answer.
+
+## Implementation
+
+Now that we understand what COSP is, let's see how we can implement it in instructor. Note that here we'll measure repetitiveness using exact string match for simplicity.
+
+!!! Info "Quick Note"
+
+    Make sure to install `numpy` and `scikit-learn` before running the code, we need to use these two libraries to compute the cosine similarity
```

Comment:
```suggestion
    Make sure to `pip install numpy scikit-learn` as we will need them for cosine computations
```

can also implement a simple costine function in a list comprehension to avoid a install